Adding support for FAISS JNI

luyuncheng opened the pull request:
*As Issue #225 mentioned. It is possible to support faiss as future.

But i do not implement the codec services, because i want to ask for your opinions about this part.

I prefer to use the same interface(**`saveIndex`, `init`, `queryIndex`**) as same as nmslib's code with the same input parameter, this would i have more compatibility for knn plugin. so i PR this using the same JNI interface.
i think with the same interface we can support nmslib and faiss in knn plugin at the same time.

*Description of changes:*
1. This PR is support `Faiss` index and Query with the same interface in jni code.
2. Add `index.knn.knnEngine` to distinguish faiss and nmslib.
3. When Write an Index in DocValuesConsumer, using `index.knn.knnEngine` to load the explicit Library.
4. When Query an Index in KNNWeigh, using `index.knn.knnEngine` to load the explicit Library.
5. In `KNNJNIFaissTests` tests, it can do the same logic as `KNNIndex` and `KNNJNITests` 
6. It almost PASS all `integTest` when using Faiss as Library
7. We can modified interface code as less as mush.
8. To build this, we need to [`build faiss`](https://github.com/facebookresearch/faiss/blob/master/INSTALL.md) firstly which depends on some external Libs like OpenMP,BLAS etc. 

I think this PR, there is a lot of details need to be discussion. 
PS. Is there any possible to open a faiss branch?

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Commit:
1. New Branch Support Faiss
2. Right Now Only Support JNI

Commit:
1. add submmodule faiss

Commit:
1. External Submodule

luyuncheng committed:
Update CMakeLists.txt

modified make without clang

luyuncheng committed:
Update CMakeLists.txt

modified OPENMP build in CMakeLists.txt

Commit:
1. add knnEngine as a Settings
2. When Write index and query index, we use knnEngine to verify search in faiss or nmslib

Commit:
1. Modified some FIXME

luyuncheng committed:
Merge pull request #1 from luyuncheng/faiss_withIT

Add index.knn.knnEngine in the ES Settings. When Write a Segment OR Query Index. It would use this settings to find the knn lib

Commit:
1. gradle build jni and jniFaiss
2. almost integTest and unitest PASSED

luyuncheng committed:
Merge pull request #2 from luyuncheng/faiss_withIT

Gradle build jni and jniFaiss

jmazanec15 commented:
Hi @luyuncheng , thanks for making this PR! This is a very big change, so it will take some time to review. I will start looking at it in January.

jmazanec15 commented on jniFaiss/CMakeLists.txt:
```diff
@@ -0,0 +1,144 @@
+cmake_minimum_required(VERSION 2.8)
```
Design question: Should we have separate JNI libraries for FAISS and nmslib, or should they be contained in one?

luyuncheng replied:
i am not sure which is better.
At this, i separate libraries just to elaborate this `faiss engine` can work with knn-plugin.
may be one jni interface can make jni code more simple to maintain

vamshin replied:
I feel having seperate JNI would be more cleaner and easy to abstract out the underlying business logic to dedicated files.  I like the current approach.

jmazanec15 replied:
I think having 2 separate libraries is okay.

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNIndexCache.java:
```diff
@@ -151,12 +153,22 @@ private void onRemoval(RemovalNotification<String, KNNIndexCacheEntry> removalNo
      */
     public KNNIndex getIndex(String key, final String indexName) {
         try {
+            //TODO if Type Not consistent
```
What do you mean by this?

luyuncheng replied:
when cached an Index, should we verify which engine for this index, and then load the exactly library?

jmazanec15 replied:
I see. I think we should add a different file extension for faiss graphs as opposed to nmslib graphs. Maybe <filename>.faiss_hnsw.

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNSettings.java:
```diff
@@ -83,6 +85,7 @@
     /**
      * Default setting values
      */
+    public static final String INDEX_KNN_DEFAULT_ENGINE = "Faiss"; // nmslib, faiss
```
I think we should keep nmslib as the default engine.

luyuncheng replied:
Yes, you are right. this is just for integration test can cover the `faiss` engine code.

jmazanec15 reviewed (commented):
I think we will review the PR first and then check into a develop branch for further testing.

Maintaining the same interface makes sense.

I added a preliminary review. I will continue to review as I think more about it.

A couple comments:
1. For FAISS submodule, we should align it with a release. It appears to be aligned to an arbitrary commit
2. In general, I do not think it is a good idea to check a binary in (libfaiss.a). What is the reason for doing this?

I need to think more on best way to support 2 libraries from the user perspective. I will get back on this.

luyuncheng committed:
Merge pull request #6 from opendistro-for-elasticsearch/master

upstream master merge into local master

Commit:
1. FIXED SpaceType Get From IndexSettings may be EMPTY when a anonymous field create.

luyuncheng committed:
Update CI.yml

Commit:
1. Add CI branches for faiss

Commit:
1. Update Cmake Files

Commit:
OFF GPU

Commit:
OFF PYTHON

Commit:
delete COMPILER

vamshin commented on build.gradle:
```diff
@@ -120,21 +120,37 @@ loggerUsageCheck.enabled = false
+    dependsOn cmakeJniLib
+    dependsOn buildJniNmsLib
+    dependsOn buildJniFaissLib
+}
```
You could remove  ```dependsOn cmakeJniLib``` and just rely on ```dependsOn buildJniNmsLib``` ```dependsOn buildJniFaissLib``` as they internally depends on ```cmakeJniNmsLib``` and ```cmakeJniFaissLib```

luyuncheng replied:
Thanks, I fixed it

vamshin commented on jniFaiss/include/com_amazon_opendistroforelasticsearch_knn_index_faiss_KNNFIndex.h:
```diff
@@ -0,0 +1,53 @@
+ * Method:    saveIndex
+ * Signature: ([I[[FLjava/lang/String;[Ljava/lang/String;Ljava/lang/String;)V
+ */
+JNIEXPORT void JNICALL Java_com_amazon_opendistroforelasticsearch_knn_index_faiss_KNNFIndex_saveIndex
```
Is faiss library backward compatible? 
We should have faiss library version in the function names to support backward compatibility issues if any arises in the future? You could refer nmslib apis.

luyuncheng replied:
Yes, your are right. i will add the version number into it.

vamshin commented on jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_KNNFIndex.cpp:
```diff
@@ -0,0 +1,272 @@
+ * When autoclose class do close, then delete the pointer
+ * Method GC pointer
+ */
+JNIEXPORT void JNICALL Java_com_amazon_opendistroforelasticsearch_knn_index_faiss_KNNFIndex_gc
```
We should run memory leak tests for faiss jni code. @jmazanec15 ran some memory leak tests for nmslib, we could probably ran it on the faiss branch once we create. Lets see if we can test it as part of this PR or atleast we can create an issue to track all TODOs later.

luyuncheng replied:
is there any tools to detect memory leak in JNI code? 
when i submit the faiss jni code, i have been writed a demo with `valgrind` to detect memory leak. but i do not know how to detect it from java test code.

vamshin replied:
Good point. We used ```valgrind``` to detect memory leaks. Last time I remember we build C++ class to call the functions directly and then build the artifact for running ```valgrind```. We could probably help here.

vamshin commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/util/NmsLibVersion.java:
```diff
@@ -25,9 +28,17 @@
             return "KNNIndexV2_0_6";
         }
+    },
+    VFaiss("Faiss") {
```
Having Faiss part of ```NmsLibVersion``` class would add confusion. We should probably have dedicated enum class for Faiss versions? Intension of this class is to hold different versions of same library and refer them from this class incase we happen to maintain more than one version of the library because of compatability issues.

luyuncheng replied:
Nice idea! i'll do it

luyuncheng replied:
How about make 
nmslib: KNNIndex_NMSLIB_V2_0_6, 
faiss: KNNIndex_FAISS_V1_6_4 
like this:
```
public enum NmsLibVersion {
    VNMSLIB_208("NMSLIB_208"){
        @Override
        public String indexLibraryVersion() {
            return "KNNIndex_NMSLIB_V2_0_8";
        }
    },
    VFAISS_164("FAISS_164") {
        @Override
        public String indexLibraryVersion() {
            return "KNNIndex_FAISS_V1_6_4";
        }
    };
}
```
or do you have any idea to distinguish the name between nmslib and faiss library.

vamshin replied:
We could have new enum class for FAISS that just holds FAISS versions.  Example

```
public enum FAISSLibVersion {

    /**
     * Latest available faiss version
     */
    V_164("164"){
        @Override
        public String indexLibraryVersion() {
            return "KNNIndexV2_0_11";
        }
    };

    public static final FAISSLibVersion LATEST = V164;

    public String buildVersion;

    FAISSLibVersion(String buildVersion) {
        this.buildVersion = buildVersion;
    }

    /**
     * FAISS library version used by the KNN codec
     * @return name
     */
    public abstract String indexLibraryVersion();
}
```

luyuncheng replied:
> We could have new enum class for FAISS that just holds FAISS versions. Example

yes, it is great. i added FAISSLibVersion in the latest commits

vamshin commented on src/test/java/com/amazon/opendistroforelasticsearch/knn/index/KNNJNIFaissTests.java:
```diff
@@ -0,0 +1,177 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class KNNJNIFaissTests extends KNNTestCase {
```
We should also have some tests that create both nmslib and faiss indices and perform search in the same test and assert right values are retrieved. This is to verify we are not mixing up between library calls.

luyuncheng replied:
yes, i would add tests for mixed engine and test with exceptions case

vamshin commented on src/main/plugin-metadata/plugin-security.policy:
```diff
@@ -1,3 +1,4 @@
 grant {
     permission java.lang.RuntimePermission "loadLibrary.KNNIndexV2_0_6";
+    permission java.lang.RuntimePermission "loadLibrary.KNNIndexVFaiss";
```
We should add version information similar to NMSlib to keep it consistent and provides ability to load certain version of the library.

luyuncheng replied:
How about make 
nmslib: KNNIndex_NMSLIB_V2_0_6, 
faiss: KNNIndex_FAISS_V1_6_4 
like this:
```
public enum NmsLibVersion {
    VNMSLIB_208("NMSLIB_208"){
        @Override
        public String indexLibraryVersion() {
            return "KNNIndex_NMSLIB_V2_0_8";
        }
    },
    VFAISS_164("FAISS_164") {
        @Override
        public String indexLibraryVersion() {
            return "KNNIndex_FAISS_V1_6_4";
        }
    };
}
```
or do you have any idea to distinguish the name between nmslib and faiss library.

With this, we may need to change the package 
`com.amazon.opendistroforelasticsearch.knn.index.v208` to `com.amazon.opendistroforelasticsearch.knn.index.nmslib.v208` and `com.amazon.opendistroforelasticsearch.knn.index.faiss` to `com.amazon.opendistroforelasticsearch.knn.index.faiss.v164`

Load Library:
```
    permission java.lang.RuntimePermission "loadLibrary.KNNIndex_NMSLIB_V2_0_8";
    permission java.lang.RuntimePermission "loadLibrary.KNNIndex_FAISS_V1_6_4";
```

vamshin commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/faiss/KNNFIndex.java:
```diff
@@ -0,0 +1,138 @@
+package com.amazon.opendistroforelasticsearch.knn.index.faiss;
```
package name should have version information similar to nmslib.

luyuncheng replied:
in the latest commits, i add `faiss` version into package and function names.

vamshin commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/codec/KNN80Codec/KNN80DocValuesConsumer.java:
```diff
@@ -98,16 +99,29 @@ public void addKNNBinaryField(FieldInfo field, DocValuesProducer valuesProducer)
-                            return null;
+
+            if(knnEngine.contains(NmsLibVersion.VFaiss.getBuildVersion())) {
+                AccessController.doPrivileged(
+                        new PrivilegedAction<Void>() {
+                            public Void run() {
+                                KNNFIndex.saveIndex(pair.docs, pair.vectors, tempIndexPath, algoParams, spaceType);
+                                return null;
+                            }
+                        }
+                );
+            } else {
+                AccessController.doPrivileged(
+                        new PrivilegedAction<Void>() {
+                            public Void run() {
+                                KNNIndex.saveIndex(pair.docs, pair.vectors, tempIndexPath, algoParams, spaceType);
+                                return null;
+                            }
                         }
-                    }
-            );
+                );
+            }
```
How about we move if else check inside the ```AccessController.doPrivileged```. Something like below example

```
                AccessController.doPrivileged(
                        new PrivilegedAction<Void>() {
                            public Void run() {
                                if(knnindex)
                                      KNNIndex.saveIndex(pair.docs, pair.vectors, tempIndexPath, algoParams, spaceType);
                                if(faissindex)
                                     KNNFIndex.saveIndex(pair.docs, pair.vectors, tempIndexPath, algoParams, spaceType);      
                               return null;
                            }
      ```

vamshin commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNWeight.java:
```diff
@@ -97,11 +100,21 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
-            );
+            final KNNQueryResult[] results;
+
+            if (fieldAttributes.containsValue(NmsLibVersion.V206.getBuildVersion())) {
```
How about we pull out the methods in KNNFIndex and KNNIndex to interface so that we could refer with one object to keep it more cleaner? Something like

```
KNNIndex index;
if (fieldAttributes.containsValue(NmsLibVersion.V206.getBuildVersion())) {
       index = knnIndexCache.getIndex(indexPath.toString(), knnQuery.getIndexName());
} else {
       index = knnIndexCache.getFIndex(indexPath.toString(), knnQuery.getIndexName());
}
results = index.queryIndex(knnQuery.getQueryVector(), knnQuery.getK());



```

luyuncheng replied:
this is may make the log of changes in KNNIndex. In KNNIndex class, it would load the specific library, if we want to pull out the methods `getIndex`, we need an abstraction of KNNIndex and verify the engine we are using.
Let me think about it how to do it.

luyuncheng replied:
> How about we pull out the methods in KNNFIndex and KNNIndex to interface so that we could refer with one object to keep it more cleaner? Something like

in the latest commits, I used KNNIndex as abstract class. `KNNFaissIndex` and `KNNNmsLibIndex` extends the KNNINdex. 
`KNNIndexCacheEntry` only save the KNNIndex.

if you have any other idea, pls let me know

vamshin commented on jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_KNNFIndex.cpp:
```diff
@@ -0,0 +1,272 @@
+#include "com_amazon_opendistroforelasticsearch_knn_index_faiss_KNNFIndex.h"
```
This class has core logic and would need some insights into way faiss library operates to review the code. I plan to visit this code once I have understanding of faiss interface.

vamshin reviewed (commented):
Few questions

1. Does faiss support all the library functions nmslib supports. We are primarily focusing on
L2, cosine, L1, negative dot product. Accordingly we may need to have relevant validations and throw the exceptions with proper message.

2. Can we confirm From Elasticsearch point of view following work? Manual test is fine for now and ideally we would prefer to have integration tests. 
    1. A nmslib index should be possible to reindex into faiss index and vice versa
    2. Forcemerge calls work on faiss index (recall should be about same)
    3. snapshot/restore work

luyuncheng committed:
Update CMakeLists.txt

update CMakeLists.txt

luyuncheng committed:
Update CMakeLists.txt

luyuncheng committed:
Update build.gradle

Fixed depends On in gradle

luyuncheng committed:
Update com_amazon_opendistroforelasticsearch_knn_index_faiss_KNNFIndex.cpp

luyuncheng committed:
Update CMakeLists.txt

luyuncheng committed:
Merge branch 'faiss' into faiss_dev

luyuncheng committed:
Update KNNCircuitBreakerIT.java

luyuncheng committed:
Update KNNCircuitBreakerIT.java

luyuncheng committed:
Update KNNCircuitBreakerIT.java

luyuncheng committed:
Update KNNCodecTestCase.java

luyuncheng committed:
Merge pull request #7 from luyuncheng/faiss_dev

Fixed using CI run  integration tests

luyuncheng committed:
Merge branch 'master' into faiss

luyuncheng committed:
Update test-workflow.yml

luyuncheng committed:
Merge pull request #8 from luyuncheng/faiss_dev

Update test-workflow.yml

Commit:
Merge branch 'master' into faiss_dev

# Conflicts:
#	src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNIndexCache.java
#	src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorFieldMapper.java
#	src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNWeight.java
#	src/main/java/com/amazon/opendistroforelasticsearch/knn/index/util/NmsLibVersion.java
#	src/main/plugin-metadata/plugin-security.policy
#	src/test/java/com/amazon/opendistroforelasticsearch/knn/index/codec/KNNCodecTestCase.java

Commit:
Merge Master Into Faiss

Commit:
Merge Master Into Faiss

Commit:
Merge Master Into Faiss

luyuncheng committed:
Merge branch 'faiss' into faiss_dev

luyuncheng committed:
Merge pull request #10 from luyuncheng/faiss_dev

Merge Upstream Maser into faiss

luyuncheng commented:
> 2\. Can we confirm From Elasticsearch point of view following work? Manual test is fine for now and ideally we would prefer to have integration tests. @vamshin 

I added some code for the integration tests in CI. and it pass all the tests. 
the default faiss library build with CPU part only. which reference from:[ general-compilation-instructions ](https://github.com/facebookresearch/faiss/blob/master/INSTALL.md#general-compilation-instructions)

Next, i'll do following things:
- [x] add faiss library version in the function names
- [x] add some tests for both faiss and nmslib library

add faiss library version in the function names with following versions:

```
public enum NmsLibVersion {
    VNMSLIB_208("NMSLIB_208"){
        @Override
        public String indexLibraryVersion() {
            return "KNNIndex_NMSLIB_V2_0_8";
        }
    },
    VFAISS_164("FAISS_164") {
        @Override
        public String indexLibraryVersion() {
            return "KNNIndex_FAISS_V1_6_4";
        }
    };
}
```

Commit:
move IndexClass check inside the AccessController.doPrivileged

Commit:
Add Tests for Circuit breaker test with different engine

Commit:
1. Add NmsLib Version And Faiss Version into KNN Plugin

luyuncheng committed:
Update CMakeLists.txt

update Lib Version

luyuncheng committed:
Update CMakeLists.txt

luyuncheng committed:
Update CMakeLists.txt

luyuncheng committed:
Update KNNWeight.java

luyuncheng committed:
Update CMakeLists.txt

Commit:
1. Add NmsLib Version And Faiss Version into KNN Plugin

Commit:
Merge remote-tracking branch 'origin/faiss_dev_libversion' into faiss_dev_libversion

Commit:
1. Add NmsLib Version And Faiss Version into KNN Plugin

luyuncheng committed:
Merge pull request #11 from luyuncheng/faiss_dev_libversion

Add Faiss and nmslib LibVersion into knn plugin

luyuncheng committed:
Add NmsLib Version And Faiss Version into KNN Plugin

1. Add NmsLib Version And Faiss Version into KNN Plugin

jmazanec15 commented on .github/workflows/CI.yml:
```diff
@@ -4,6 +4,7 @@ on:
     branches:
       - master
       - opendistro-*
+      - faiss*
```
I dont think we need to add this branch to CI action, given that it is a development branch. Please remove. Once review finishes, we will check into faiss branch and develop/test from there. Then, once we are ready, we can create another PR to master.

luyuncheng replied:
ok，i'll fix it

jmazanec15 commented on .github/workflows/test-workflow.yml:
```diff
@@ -5,10 +5,12 @@ on:
     branches:
       - master
       - opendistro-*
+      - faiss*
```
Same as above. Please remove.

luyuncheng committed:
Merge pull request #13 from opendistro-for-elasticsearch/master

Update Upstream Master

Commit:
1. Add NmsLib Version And Faiss Version into KNN Plugin

Commit:
1. Merge remote-tracking branch 'origin/master' into faiss_dev
2. Make Faiss Version into code
3. Update Hnswlib to 2.0.10

# Conflicts:
#	jni/CMakeLists.txt
#	jni/include/com_amazon_opendistroforelasticsearch_knn_index_nmslib_v208_KNNIndex.h
#	jni/include/com_amazon_opendistroforelasticsearch_knn_index_v2011_KNNIndex.h
#	jni/include/com_amazon_opendistroforelasticsearch_knn_index_v208_KNNIndex.h
#	jni/src/com_amazon_opendistroforelasticsearch_knn_index_nmslib_v208_KNNIndex.cpp
#	jni/src/com_amazon_opendistroforelasticsearch_knn_index_v2011_KNNIndex.cpp
#	jni/src/com_amazon_opendistroforelasticsearch_knn_index_v208_KNNIndex.cpp
#	src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNIndexCache.java
#	src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNIndexShard.java
#	src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNWeight.java
#	src/main/java/com/amazon/opendistroforelasticsearch/knn/index/codec/KNN80Codec/KNN80DocValuesConsumer.java
#	src/main/java/com/amazon/opendistroforelasticsearch/knn/index/nmslib/v208/KNNIndex.java
#	src/main/java/com/amazon/opendistroforelasticsearch/knn/index/util/NmsLibVersion.java
#	src/main/java/com/amazon/opendistroforelasticsearch/knn/index/v2011/KNNIndex.java
#	src/main/java/com/amazon/opendistroforelasticsearch/knn/index/v208/KNNIndex.java
#	src/main/plugin-metadata/plugin-security.policy
#	src/test/java/com/amazon/opendistroforelasticsearch/knn/index/KNNIndexCacheTests.java
#	src/test/java/com/amazon/opendistroforelasticsearch/knn/index/KNNJNITests.java

jmazanec15 commented on jniFaiss/CMakeLists.txt:
```diff
@@ -0,0 +1,108 @@
+cmake_minimum_required(VERSION 2.8)
+
+project(KNNIndex_FAISS)
```
I think this might be better for CMakeLists.txt: https://gist.github.com/jmazanec15/da1e68aed9073c7d4706597ba8ef2087.

luyuncheng replied:
nice code! i'll use this one

jmazanec15 commented on jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.cpp:
```diff
@@ -0,0 +1,272 @@
+
+std::unordered_map<string, faiss::MetricType> mapMetric = {
+        {"l2", faiss::METRIC_L2},
+        {"innerproduct", faiss::METRIC_INNER_PRODUCT}
```
Doesn't FAISS also have cosine space?

luyuncheng replied:
In Doc [Faiss-indexes](https://github.com/facebookresearch/faiss/wiki/Faiss-indexes) , i added inner product, which means If we have normalized vectors, the inner product becomes cosine similarity. 
And there is some issues talk about this: https://github.com/facebookresearch/faiss/issues/593

jmazanec15 replied:
Oh interesting. That makes sense.

jmazanec15 commented on jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.cpp:
```diff
@@ -0,0 +1,272 @@
+
+extern "C"
+
+struct JavaException {
```
Can you add a TODO for future to abstract out common logic from JNI levels? This does not have to be completed in this PR, but we should do it in the future.

jmazanec15 commented on jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.cpp:
```diff
@@ -0,0 +1,272 @@
+	vector<float>   dataset;
+	vector<string> paramsList;
+	string indexDescription = "HNSW32";
+	faiss::MetricType metric = faiss::METRIC_L2;
```
I think instead of setting L2 as default, we should just make sure that spaceType is passed in and is valid. Setting L2 as default here could lead to silent failures.

luyuncheng replied:
L72 just init a default value. at L121 i verify the Metric Space:
```
if(mapMetric.find(spaceTypeString) != mapMetric.end()) {
    metric = mapMetric[spaceTypeString];
}
```

I have a question: if we verify the `metricSpace` in the jni code. this may cause ES shard failed assert when an invalid space param passed in jni code.  should i throw an exception when mapMetric can not find spaceTypeString?

jmazanec15 commented on jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.cpp:
```diff
@@ -0,0 +1,272 @@
+
+		//---- indexPath
+		const char *indexString = env->GetStringUTFChars(indexPath, 0);
+        string indexPathString(indexString);
```
Indentation in this file looks off. Could you please fix?

jmazanec15 commented on jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.cpp:
```diff
@@ -0,0 +1,272 @@
+	vector<int64_t> idVector;
+	vector<float>   dataset;
+	vector<string> paramsList;
+	string indexDescription = "HNSW32";
```
Could you add a TODO here to eventually parametrize this? I think we should be able to support any Faiss index in the future.

jmazanec15 commented on jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.cpp:
```diff
@@ -0,0 +1,272 @@
+            paramsList.push_back(rawString);
+
+            int M = 32;
+            if (sscanf(rawString, "M=%d", &M) == 1) {
```
Why do we need to sscanf here?

luyuncheng replied:
> Why do we need to sscanf here?

In the KNN plugin we would make algoParam like: `M=32, efSearch=20`, As [faiss index factory docs](https://github.com/facebookresearch/faiss/wiki/The-index-factory#non-exhaustive-search-components) shows, if `M=32` we need use string: `HNSW32` for index Description. so i use sscanf to get M.

i think we need an adapter to match different knnEngine algorithms params. if you have good idea, pls let me know.

jmazanec15 replied:
Ah got it. I think makes sense.

jmazanec15 commented on jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.cpp:
```diff
@@ -0,0 +1,272 @@
+		//---- Do Index
+		//----- 1. Train
+        if(!indexWriter->is_trained) {
+			//TODO if we use like PQ, we have to train dataset
```
Yeah I was thinking about this a little bit. It seems like it might make sense to fall back to a flat index if we do not have enough points to train. But we can think about this later.

jmazanec15 commented on jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.cpp:
```diff
@@ -0,0 +1,272 @@
+{
+	//set thread 1 cause ES has Search thread
+	//TODO make it different at search and write
+//	omp_set_num_threads(1);
```
Why is this commented out?

luyuncheng replied:
i am no sure is it appropriate to set threads as 1 or  reuse the settings:
`public static final String KNN_ALGO_PARAM_INDEX_THREAD_QTY = "knn.algo_param.index_thread_qty";`

jmazanec15 replied:
I think it probably makes sense to reuse KNN_ALGO_PARAM_INDEX_THREAD_QTY.

jmazanec15 commented on src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNIndexCache.java:
```diff
@@ -320,22 +339,52 @@ public KNNIndexCacheEntry loadIndex(String indexPathUrl, String indexName) throw
         private KNNIndexCacheEntry(final KNNIndex knnIndex, final String indexPathUrl, final String esIndexName,
                                    final WatcherHandle<FileWatcher> fileWatcherHandle) {
             this.knnIndex = knnIndex;
+            this.knnFindex = null;
```
Seems to me that knnIndex and knnFindex should inherit from some kind of abstract class.

luyuncheng replied:
> I see. I think we should add a different file extension for faiss graphs as opposed to nmslib graphs. Maybe .faiss_hnsw. 

in the latest commit, i use `knnEngine` name in the index file names like:
```
String hnswFileName = String.format("%s_%s_%s%s", 
    state.segmentInfo.name, 
    knnEngine.getLatestBuildVersion(),
    field.name, 
    KNNCodecUtil.HNSW_EXTENSION);
```

> Seems to me that knnIndex and knnFindex should inherit from some kind of abstract class.

yes, you are right. let me think about how to abstract this.

luyuncheng replied:
> Seems to me that knnIndex and knnFindex should inherit from some kind of abstract class.

I used KNNIndex as abstract class. `KNNFaissIndex` and `KNNNmsLibIndex` extends the KNNINdex. 
`KNNIndexCacheEntry` only save the KNNIndex.

if you have any other idea, pls let me know

jmazanec15 replied:
I think that makes sense. Thanks for making the change!

jmazanec15 reviewed (commented):
Hi @luyuncheng, could you remove the jniFaiss/external/libfaiss.a? We don't need to ship binaries.

Also, can you update FAISS library commit to their latest release? v1.6.5. Right now, it points to a more recent commit.

Commit:
1. Update Submodule FAISS to new version
2. Remove no useful libfaiss.a
3. Update JniFaiss CmakeLists.txt

Commit:
1. Update Submodule FAISS to new version V1.6.5(HASH:88eabe9)

Commit:
1. Update Submodule FAISS to new version V1.6.5(HASH:88eabe9)
2. Remove no use comments

Commit:
1. Remove nouse CI branches

luyuncheng committed:
Merge pull request #14 from luyuncheng/faiss_dev

Add NmsLib Version And Faiss Version into KNN Plugin

Commit:
1. Fixed Indentation

Commit:
1. Make KNNIndex as abstract class
2. Add KNNNmsLibIndex and KNNFaissIndex inherit the KNNIndex

Commit:
1. Make KNNIndex as abstract class
2. Add KNNNmsLibIndex and KNNFaissIndex inherit the KNNIndex
3. Move JNITests into JNINmsLibTests
4. Add New JNITests which would mixed knnEngine, and test with exception

luyuncheng committed:
Merge pull request #15 from luyuncheng/opendistro-faiss_dev

Make KNNIndex as abstract class

jmazanec15 reviewed (commented):
Thanks for making this change @luyuncheng!

I think at this point we can check it into a development branch "faiss-support" and I will start working on it/testing it as well.

I created "faiss-support". Could you merge in all the changes to that branch so that we can merge this change in?

luyuncheng commented:
> Thanks for making this change @luyuncheng!
> 
> I think at this point we can check it into a development branch "faiss-support" and I will start working on it/testing it as well.
> 
> I created "faiss-support". Could you merge in all the changes to that branch so that we can merge this change in?

i have changed the base branch from `master` to `faiss-support`.  
Thanks for reviewing this PR ! @jmazanec15 @vamshin 
Let me know if you need anything else

vamshin reviewed (approved):
LGTM! to push to dedicate branch. Once again thanks for the great work @luyuncheng.

We can continue development/testing on top of this code in dedicated branch.
